p2p/pkg/rpc/bidder/service.go

```diff
@@ -316,3 +319,204 @@ func (s *Service) Withdraw(
 
 	return nil, status.Errorf(codes.Internal, "missing log for withdrawal")
 }
+
+func (s *Service) AutoDeposit(
+	ctx context.Context,
+	r *bidderapiv1.DepositRequest,
+) (*bidderapiv1.AutoDepositResponse, error) {
+	err := s.validator.Validate(r)
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "validating auto deposit request: %v", err)
+	}
+
+	if s.autoDepositTracker.IsWorking() {
+		return nil, status.Error(codes.FailedPrecondition, "auto deposit is already running")
+	}
+
+	currentWindow, err := s.blockTrackerContract.GetCurrentWindow()
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "getting current window: %v", err)
+	}
+
+	windowToDeposit, err := s.calculateWindowToDeposit(ctx, r, currentWindow.Uint64())
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "calculating window to deposit: %v", err)
+	}
+
+	amount, success := big.NewInt(0).SetString(r.Amount, 10)
+	if !success {
+		return nil, status.Errorf(codes.InvalidArgument, "parsing amount: %v", r.Amount)
+	}
+
+	err = s.autoDepositTracker.Start(ctx, windowToDeposit, amount)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "starting auto deposit: %v", err)
+	}
+
+	s.logger.Error(
+		"autodeposit enabled",
+		"window", windowToDeposit,
+		"amount", amount.String(),
+	)
+
+	return &bidderapiv1.AutoDepositResponse{
+		StartBlockNumber: wrapperspb.UInt64(windowToDeposit.Uint64()),
+		AmountPerWindow:  amount.String(),
+	}, nil
+}
+
+func (s *Service) CancelAutoDeposit(
+	ctx context.Context,
+	r *bidderapiv1.CancelAutoDepositRequest,
+) (*bidderapiv1.CancelAutoDepositResponse, error) {
+	if !s.autoDepositTracker.IsWorking() {
+		return nil, status.Error(codes.FailedPrecondition, "auto deposit is not running")
+	}
+	windows, err := s.autoDepositTracker.Stop()
+	if err != nil {
+		return nil, status.Errorf(codes.FailedPrecondition, "cancel auto deposit: %v", err)
+	}
+	if r.Withdraw {
+		go func() {
+			ticker := time.NewTicker(5 * time.Second)
+			defer ticker.Stop()
+			for range ticker.C {
+				currentWindow, err := s.blockTrackerContract.GetCurrentWindow()
+				if err != nil {
+					s.logger.Error("getting current window", "error", err)
+					continue
+				}
+				doWithdraw := true
+				for _, w := range windows {
+					if w.Uint64() >= currentWindow.Uint64() {
+						doWithdraw = false
```

Comment:
It would probably be good to add a log here indicating that we are waiting for the correct window, so it will be clearer to the user what is happening